### ISSUE TITLE
view_slurp and view_iterator leak memory

### ISSUE THREAD

**Issue by ivulfson** [mnunberg/perl-Couchbase-Client#34]
Looks like `view_slurp` just never releases memory.  Subsequent calls to `view_slurp` eat more memory, and more memory, and nom nom nom.

The `view_iterator` is a lot better, but still leaks about 44kb on every call.

```perl
while (1) {
        my $iter   = $CB-view_iterator("$design/$view", stale => 'ok', limit => 1);
        while (my $row = $iter->next) {
            # do something
        }   
}
```

Running this will eventually cause it to run out of ram.  Even this:

```perl
while (1) {
        my $iter   = $CB-view_iterator("$design/$view", stale => 'ok', limit => 1);
}
```
Using `stop` or `undef` on `$iter` doen't help.

**Comment by mnunberg** [comment]
Time to rip up this view implementation and use the new one in libcouchbase...

**Comment by ivulfson** [comment]
I love it!

**Comment by mnunberg** [comment]
Please see the `views-wip` branch. I don't know how done it is, but I believe it fixes the memory leak.. by removing large swathes of cruft.

Note that when running it I did see some odd memory usage for the first 5-10 minutes or so, but I assume this is just Perl adjusting its memory pools. Usage for the standard "beer-sample" view (in a busy loop) seems to stabilize at around 20M after that.

**Comment by ivulfson** [comment]
cpan2rpm coughs on it, it seems.


```
...
gcc -c  -I/tmp/lcb/include -D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic   -DVERSION=\"2.0.2\" -DXS_VERSION=\"2.0.2\" -o xs/Couchbase.o -fPIC "-I/usr/lib64/perl5/CORE"   xs/Couchbase.c
In file included from xs/Couchbase.xs:1:
xs/perl-couchbase.h:218:34: error: perl-couchbase-couch.h: No such file or directory
xs/Couchbase.xs: In function ‘PLCB_construct’:
xs/Couchbase.xs:86: error: ‘PLCB_VIEWHANDLE_CLASS’ undeclared (first use in this function)
xs/Couchbase.xs:86: error: (Each undeclared identifier is reported only once
xs/Couchbase.xs:86: error: for each function it appears in.)
xs/Couchbase.c: In function ‘XS_Couchbase__viewhandle_new’:
xs/Couchbase.c:1455: warning: implicit declaration of function ‘PLCB__viewhandle_new’
xs/Couchbase.c:1455: warning: assignment makes pointer from integer without a cast
xs/Couchbase.c: In function ‘XS_Couchbase__viewhandle_fetch’:
xs/Couchbase.c:1476: warning: implicit declaration of function ‘PLCB__viewhandle_fetch’
make: *** [xs/Couchbase.o] Error 1
mk_rpm(): Inappropriate ioctl for device at /home/ivulfson/cpan2rpm line 1012.
-- Done --
```

**Comment by mnunberg** [comment]
Seems i forgot to commit a file..

**Comment by mnunberg** [comment]
Try it again now..

**Comment by ivulfson** [comment]
```
xs/operations.c: In function ‘PLCB_op_observe’:
xs/operations.c:139: warning: passing argument 2 of ‘key_from_so’ from incompatible pointer type
xs/operations.c:17: note: expected ‘struct lcb_CMDBASE *’ but argument is of type ‘struct lcb_CMDOBSERVE *’
xs/operations.c: In function ‘PLCB_op_endure’:
xs/operations.c:170: warning: passing argument 2 of ‘key_from_so’ from incompatible pointer type
xs/operations.c:17: note: expected ‘struct lcb_CMDBASE *’ but argument is of type ‘struct lcb_CMDENDURE *’
xs/operations.c:171: warning: passing argument 3 of ‘PLCB_args_unlock’ from incompatible pointer type
xs/plcb-args.h:60: note: expected ‘struct lcb_CMDUNLOCK *’ but argument is of type ‘struct lcb_CMDENDURE *’
xs/operations.c:172: warning: passing argument 2 of ‘mctx->addcmd’ from incompatible pointer type
xs/operations.c:172: note: expected ‘const struct lcb_CMDBASE *’ but argument is of type ‘struct lcb_CMDENDURE *’
gcc -c  -I/tmp/lcb/include -D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic   -DVERSION=\"2.0.2\" -DXS_VERSION=\"2.0.2\" -o xs/opcontext.o -fPIC "-I/usr/lib64/perl5/CORE"   xs/opcontext.c
Running Mkbootstrap for Couchbase ()
chmod 644 Couchbase.bs
rm -f blib/arch/auto/Couchbase/Couchbase.so
gcc  -shared -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic xs/Couchbase.o xs/BucketConfig.o xs/IO.o xs/args.o xs/async.o xs/callbacks.o xs/constants.o xs/convert.o xs/operations.o xs/opcontext.o xs/viewrow.o  -o blib/arch/auto/Couchbase/Couchbase.so 	\
	   -L/tmp/lcb/lib -lcouchbase  	\
	  
gcc: xs/viewrow.o: No such file or directory
make: *** [blib/arch/auto/Couchbase/Couchbase.so] Error 1
```

**Comment by mnunberg** [comment]
ffs.. but I've verified the most recent commit (in which i apparently forgot _another_ file) should work..

**Comment by ivulfson** [comment]
Yes, it builds now.  There are a couple of bugs in cpan2rpm, I may deal with those on Monday.

There are still a few warnings.  I especially love the one about `meth`:

```
gcc -c  -I/tmp/lcb/include -D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic   -DVERSION=\"2.0.2\" -DXS_VERSION=\"2.0.2\" -o xs/Couchbase.o -fPIC "-I/usr/lib64/perl5/CORE"   xs/Couchbase.c
gcc -c  -I/tmp/lcb/include -D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic   -DVERSION=\"2.0.2\" -DXS_VERSION=\"2.0.2\" -o xs/BucketConfig.o -fPIC "-I/usr/lib64/perl5/CORE"   xs/BucketConfig.c
gcc -c  -I/tmp/lcb/include -D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic   -DVERSION=\"2.0.2\" -DXS_VERSION=\"2.0.2\" -o xs/IO.o -fPIC "-I/usr/lib64/perl5/CORE"   xs/IO.c
gcc -c  -I/tmp/lcb/include -D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic   -DVERSION=\"2.0.2\" -DXS_VERSION=\"2.0.2\" -o xs/args.o -fPIC "-I/usr/lib64/perl5/CORE"   xs/args.c
gcc -c  -I/tmp/lcb/include -D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic   -DVERSION=\"2.0.2\" -DXS_VERSION=\"2.0.2\" -o xs/async.o -fPIC "-I/usr/lib64/perl5/CORE"   xs/async.c
gcc -c  -I/tmp/lcb/include -D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic   -DVERSION=\"2.0.2\" -DXS_VERSION=\"2.0.2\" -o xs/callbacks.o -fPIC "-I/usr/lib64/perl5/CORE"   xs/callbacks.c
gcc -c  -I/tmp/lcb/include -D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic   -DVERSION=\"2.0.2\" -DXS_VERSION=\"2.0.2\" -o xs/constants.o -fPIC "-I/usr/lib64/perl5/CORE"   xs/constants.c
gcc -c  -I/tmp/lcb/include -D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic   -DVERSION=\"2.0.2\" -DXS_VERSION=\"2.0.2\" -o xs/convert.o -fPIC "-I/usr/lib64/perl5/CORE"   xs/convert.c
gcc -c  -I/tmp/lcb/include -D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic   -DVERSION=\"2.0.2\" -DXS_VERSION=\"2.0.2\" -o xs/operations.o -fPIC "-I/usr/lib64/perl5/CORE"   xs/operations.c
xs/operations.c: In function ‘PLCB_op_observe’:
xs/operations.c:139: warning: passing argument 2 of ‘key_from_so’ from incompatible pointer type
xs/operations.c:17: note: expected ‘struct lcb_CMDBASE *’ but argument is of type ‘struct lcb_CMDOBSERVE *’
xs/operations.c: In function ‘PLCB_op_endure’:
xs/operations.c:170: warning: passing argument 2 of ‘key_from_so’ from incompatible pointer type
xs/operations.c:17: note: expected ‘struct lcb_CMDBASE *’ but argument is of type ‘struct lcb_CMDENDURE *’
xs/operations.c:171: warning: passing argument 3 of ‘PLCB_args_unlock’ from incompatible pointer type
xs/plcb-args.h:60: note: expected ‘struct lcb_CMDUNLOCK *’ but argument is of type ‘struct lcb_CMDENDURE *’
xs/operations.c:172: warning: passing argument 2 of ‘mctx->addcmd’ from incompatible pointer type
xs/operations.c:172: note: expected ‘const struct lcb_CMDBASE *’ but argument is of type ‘struct lcb_CMDENDURE *’
gcc -c  -I/tmp/lcb/include -D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic   -DVERSION=\"2.0.2\" -DXS_VERSION=\"2.0.2\" -o xs/opcontext.o -fPIC "-I/usr/lib64/perl5/CORE"   xs/opcontext.c
gcc -c  -I/tmp/lcb/include -D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic   -DVERSION=\"2.0.2\" -DXS_VERSION=\"2.0.2\" -o xs/viewrow.o -fPIC "-I/usr/lib64/perl5/CORE"   xs/viewrow.c
xs/viewrow.c: In function ‘viewrow_callback’:
xs/viewrow.c:99: warning: value computed is not used
xs/viewrow.c:100: warning: value computed is not used
xs/viewrow.c:101: warning: value computed is not used
xs/viewrow.c:102: warning: value computed is not used
xs/viewrow.c:105: warning: value computed is not used
xs/viewrow.c:73: warning: unused variable ‘meth’
Running Mkbootstrap for Couchbase ()
chmod 644 Couchbase.bs
rm -f blib/arch/auto/Couchbase/Couchbase.so
gcc  -shared -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector --param=ssp-buffer-size=4 -m64 -mtune=generic xs/Couchbase.o xs/BucketConfig.o xs/IO.o xs/args.o xs/async.o xs/callbacks.o xs/constants.o xs/convert.o xs/operations.o xs/opcontext.o xs/viewrow.o  -o blib/arch/auto/Couchbase/Couchbase.so 	\
	   -L/tmp/lcb/lib -lcouchbase  	\
	  
```

**Comment by mnunberg** [comment]
It would seem my local compilation flags are laxer -- I can only get the one about `meth` to even show. Most likely hv_stores is declared with `__attribute__((warn_unused_result))` on the Perl you're using. I'm using 5.20.1 from Homebrew.

But does this fix the leak for you? I plan to add N1QL support and "spatial" support (all using more or less the same API pattern, passed down from the C core).

**Comment by ivulfson** [comment]
I haven't tested it yet, because I can't install the RPM just yet.  I did patch cpan2rpm though (see https://github.com/ekkis/cpan2rpm/pull/1), so now it actually builds perl-Couchbase.

I'm compiling on RHEL6 64-bit, gods help me.  This is for a production environment, and sadly we are still years behind, although going up to Centos 7 some time this year, maybe.

**Comment by ivulfson** [comment]
Alright, just tested it.

First thing, can't call `view_iterator` with an aref `[ $design, $view ]`, dies with `Must pass view/design ` - gotta call as `"$design/$view"`.

Second, can't call `info()` on iterator - not implemented in `Couchbase::View::Handle`.  Calling `is_ok()` and `errstr()` on the iterator object itself works now, but this API change should probably be documented.

**Comment by ivulfson** [comment]
Does this new code need a specific version of libcouchbase-devel or libcouchbase2-core RPMs?  I was using the latest libcouchbase code from git before, and now I tried using versions 2.4.4 of the RPMs to build this view , and getting this:

```
xs/constants.c:3:32: error: libcouchbase/views.h: No such file or directory
xs/constants.c: In function ‘plcb_define_constants’:
xs/constants.c:53: error: ‘LCB_CMDVIEWQUERY_F_NOROWPARSE’ undeclared (first use in this function)
xs/constants.c:53: error: (Each undeclared identifier is reported only once
xs/constants.c:53: error: for each function it appears in.)
xs/constants.c:54: error: ‘LCB_CMDVIEWQUERY_F_INCLUDE_DOCS’ undeclared (first use in this function)
xs/constants.c:55: error: ‘LCB_CMDVIEWQUERY_F_SPATIAL’ undeclared (first use in this function)
make: *** [xs/constants.o] Error 1
```

**Comment by ivulfson** [comment]
OK, compiled against 2.4.8 libs.
Yeah, it works, but `view_iterator()` is still leaking like a sieve. :)
After about 20k loops, the process grew to ~800MB, at which point I killed it.

**Comment by mnunberg** [comment]
However I have been able to reproduce a leak on the current master branch (without the views-wip fixes)

**Comment by ivulfson** [comment]
I will test with the bucket of beer.  Mmmm beer.  If not today, then tomorrow for sure.  I was testing with a view that returned about 3.5M rows, if left unlimited.

**Comment by mnunberg** [comment]
And the leak only happens when you query that big view? Does it only happen when the full resultset (or a large part thereof) is returned, or does it happen even with a single small row. This might be something to do with the library’s internal buffers, especially if the network ends up being fast and your application ends up being slow (And you are interleaving row results with actual KV operations).

libcouchbase has some pause functionality for views. I am not sure if this will help (it’s not exposed in any event, but I may end up exposing it..)

**Comment by ivulfson** [comment]
The latest pull from views-wip causes this error with cpan2rpm:

```
Error: Package: perl-Couchbase-2.0.2-1.x86_64 (/perl-Couchbase-2.0.2-1.x86_64)
           Requires: perl(Couchbase::HTTPDocument)
```
I had to add `--provides 'perl(Couchbase::HTTPDocument)'` to cpan2rpm command line args.

**Comment by ivulfson** [comment]
OK, no memory leak with the `beer/brewery_beers` view.

**Comment by ivulfson** [comment]
Running with a large bucket now, pulling in 3.5M rows per iteration, and it's slow but stable. Changed it to limit at 10k, and let it run for a while.  Memory usage stable.  Went up to 100k, good as well.

Here's the interesting part.  Using `view_slurp()` with 3.5M rows takes up about a gigabyte of RAM.  Using `view_iterator()` with same 3.5M rows takes up under 300MB of RAM.  They both finish in about the same time, though.

**Comment by mnunberg** [comment]
Well, view_slurp has to buffer the entire contents of the view before it returns it to you :) — expect it to consume quite a lot of RAM. view_iterator is the preferred function to use as it will return rows incrementally and only do network fetches once no more rows are local anymore.

I’m curious as to how slow it is compared to the older version — (and whether this improves when using a non-threaded Perl).

**Comment by ivulfson** [comment]
OK, 100 iterations using your script you posted in #issuecomment-89813216, using the beer sample.

libcouchbase 2.4.4, perl-Couchbase 2.0.1 - 41 seconds
libcouchbase 2.4.8, perl-Couchbase 2.0.2 views-wip - 42 seconds

So, never mind the couple of orders of magnitude. :)

**Comment by ivulfson** [comment]
OK, checked it with 3.5M rows view.  Same time for 2.0.1 and views-wip.

BTW, add lib/Couchbase/HTTPDocument.pm to MANIFEST, please. :)  And ship it. ;)

**Comment by ivulfson** [comment]
I just noticed that lib/Couchbase/BucketConfig.pm isn't in the manifest either.  Or none of the files in lib/Couchbase/IO.  Is that on purpose?

**Comment by ivulfson** [comment]
Oy vey.  This segfaults:

```perl
        my $iter = $bkt->view_iterator("beer/brewery_beers", %options);
        while($iter->next) {}
        $iter->stop;
```

**Comment by ivulfson** [comment]
Also, this causes a bunch of warnings:

```perl
        my $iter = $bkt->view_iterator("beer/brewery_beers", %options);
        die $iter->errstr if !$iter->is_ok;
```
```
Use of uninitialized value $num in numeric ne (!=) at /usr/local/lib64/perl5/Couchbase/Document.pm line 25.
Use of uninitialized value $num in numeric eq (==) at /usr/local/lib64/perl5/Couchbase/Document.pm line 27.
```

**Comment by mnunberg** [comment]
The lack of the IO.pm may be intentional. That's intended for async support, which I have not yet decided what to do with. I'll try to add the other stuff to the Manifest.

See if you can pull the latest commit -- I did some additional improvements in order to make the tests pass again - and thus in my local HEAD, I'm not seeing the behavior you describe (though it can very well exist in your current branch)

**Comment by mnunberg** [comment]
Indeed, it seems older Perls break in this case. Fixed in latest commit.

**Comment by ivulfson** [comment]
OK, pulled the last commit.  This no longer segfaults, nor does it leak memory:

```perl
        my $iter = $CB->view_iterator([ $design, $view ], stale => 'ok', limit => 100);
        while ($iter->next){}
        $iter->stop;
        die $iter->info->errstr if !$iter->info->is_ok;
```

This leaks memory slightly, my guess is because iterator is not exhausted, and it throws warnings:

```perl
        my $iter = $CB->view_iterator([ $design, $view ], stale => 'ok', limit => 100);
        $iter->stop;
        die $iter->info->errstr if !$iter->info->is_ok;
```

```
Use of uninitialized value $num in numeric ne (!=) at /usr/local/lib64/perl5/Couchbase/Document.pm line 25.
Use of uninitialized value $num in numeric eq (==) at /usr/local/lib64/perl5/Couchbase/Document.pm line 27.
```

This doesn't leak memory, but throws the same warnings as above:

```perl
        my $iter = $CB->view_iterator([ $design, $view ], stale => 'ok', limit => 100);
        $iter->next;
        $iter->stop;
        die $iter->info->errstr if !$iter->info->is_ok;
```

**Comment by ivulfson** [comment]
Also, it's not entirely clear what the more idiomatic (to call on my recently learned Go vocabulary) way to check for errors is: this -
```perl
die $iter->info->errstr if !$iter->info->is_ok;
```
or this
```perl
die $iter->errstr if !$iter->is_ok;
```

Both ways work.

**Comment by mnunberg** [comment]
Regarding the last comment, the `->info` object is a relic from the past (about 2 years old now), and was my first shot at designing the view API. When I redid the client for 2.0 I focused mainly on KV operations, while view operations used largely the same code: See http://search.cpan.org/~mnunberg/Couchbase-Client-2.0.0_1/lib/Couchbase/Couch/Base.pm . The ugliness is due to the fact that I pretty much had to implement a lot of the layers myself, and the interaction between various components weren't well defined. Now that views is done in the C library, all I need to do at the Perl side is collect the individual rows and handle them like that.

In short, the `->info` method is still supported to support older code, but if you now look in the source:

```
sub info {
    return $_[0];
}
```

I do hate breaking API, so either way should be fine (though I'll end up undocumenting `info` soon).

**Comment by mnunberg** [comment]
Regarding the memory leak -- the memory isn't "leaked" per se, but you would need another few iterations of the event loop for things to be cleaned up -- some of the resources created relinquish control  to the event loop, `stop` weakly signals them to terminate (and of course, not call any of our user requested callbacks).

**Comment by mnunberg** [comment]
I've now deleted the `views-wip` branch, merging it into master. Considering that N1QL "support" is being added as well, all I need is to document the new functionality and release it

**Comment by ivulfson** [comment]
Fantastic, no more `->info` then!  Love it.

No memory leaks on master branch, with either `view_slurp` or `view_iterator`.

**Comment by mnunberg** [comment]
I guess you can close this issue now.

**Comment by ivulfson** [comment]
Indeed.  Thank you very much for such quick response.  Closing.
